[Fix] Fix for CADMIN_1_5 DNS-SD commissionable device discovery issue #38538

j-ororke opened the pull request:
**Reason**
- Fix for congested commissionable discovery issue #[38538](https://github.com/project-chip/connectedhomeip/issues/38538) reported by Kishok
- Issue reported to be noticed when running in a congested testing environment when multiple test devices were in commissionable mode

**Resolution**
Improved TC_CADMIN_1_5 test reliability by changing the service discovery approach to handle discovery in the event of multiple commissionable devices. 

The test now:
1. Gathers all discoverable service records rather than just the first one as was done previously
2. Filters through all commissionable services discovered to find the correct match based on both CM value and discriminator

This makes the test more reliable in congested test environments where multiple devices might be in commissioning mode simultaneously, avoiding false failures when the correct device isn't the first one discovered.

#### Testing
- WSL

j-ororke committed:
Fix for CADMIN_1_5 test module:
- Changed to gathering all discoverable service records for a more congested test environment, filtering those service records by commissionable services, then checking those services for the correct cm and discriminator, this makes the test step more dynamic and able to find multiple commissionable services, where as previously it would only return the first record it found-

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

andy31415 commented on src/python_testing/TC_CADMIN_1_5.py:
```diff
@@ -44,13 +45,59 @@
+                d_value = service.txt_record.get('D')
+
+                # Convert to strings for comparison
+                if str(cm_value) == str(expected_cm_value) and str(d_value) == str(expected_discriminator):
```
Why do we convert to strings? Would it make sense to type safe one of them and only correct the other?

Should cm and discriminator be integers and we should try to convert to ints instead?

andy31415 replied:
```suggestion
                if cm_value == str(expected_cm_value) and d_value == str(expected_discriminator):
```

Is there any reason why cm_value or d_value would not be strings? You could also use a `get(..., 'MISSING')` as you do below.

j-ororke replied:
Using the dataclass method was definitely the way to go, the above code is no longer needed and has been removed.

andy31415 commented on src/python_testing/TC_CADMIN_1_5.py:
```diff
@@ -122,11 +169,16 @@ async def test_TC_CADMIN_1_5(self):
+        # Wait for DNS-SD advertisement with correct CM value and discriminator
+        # This will either return a valid service or assert failure
+        service = await self.wait_for_correct_cm_value(
+            expected_cm_value="2",
```
```suggestion
            expected_cm_value=2,
```

Seems better since spec seems to have these integers

j-ororke replied:
This does seem better, the DNS-SD returns the values as strings, but have changed to handle this inside the function.

andy31415 commented on src/python_testing/TC_CADMIN_1_5.py:
```diff
@@ -44,13 +45,59 @@
+            return discovery._discovered_services[mdns_discovery.MdnsServiceType.COMMISSIONABLE.value]
+        return []
+
+    async def wait_for_correct_cm_value(self, expected_cm_value, expected_discriminator, max_attempts=5, delay_sec=5):
```
```suggestion
    async def wait_for_correct_cm_value(self, expected_cm_value: int, expected_discriminator: int, max_attempts=5, delay_sec=5):
```

To avoid too much duck typing. Would this work?

j-ororke replied:
Thank you, this has been added when adding the dataclass to the code.

andy31415 commented on src/python_testing/TC_CADMIN_1_5.py:
```diff
@@ -44,13 +45,59 @@
+            if services:
+                logging.info(f"Found {len(services)} services, but none match CM={expected_cm_value}, D={expected_discriminator}")
+                for service in services:
+                    logging.info(f"  Service: CM={service.txt_record.get('CM')} (type: {type(service.txt_record.get('CM'))}), "
```
Why do we need to log the type here? Was that done for debug? It seems it should always be string...

j-ororke replied:
You are correct this is not needed any longer, I had it in there for debugging

andy31415 commented on src/python_testing/TC_CADMIN_1_5.py:
```diff
@@ -44,13 +45,59 @@
+
+            # Look through all services for a match
+            for service in services:
+                cm_value = service.txt_record.get('CM')
```
you are doing the get of CM and D in many places, as well as the format.
How about creating some `@dataclass` to and re-use and have methods on it. Like:

```python

@dataclass
class ParsedService:
    service: MdnsServiceInfo
    cm: Optional[int] = None
    d: Optional[int] = None
    
    def __post_init__(self):
        self.cm = self.service.txt_record.get('CM', None)
        self.d = self.service.txt_record.get('D', None)
        
    def __str__(self) -> str:     
        return f"Service CM={self.cm}, D={self.d}"     
```

And add any other methods after that? That may make the code clearer and the formatting of things more consistent.

j-ororke replied:
I admit I really like this approach it makes the code much cleaner by handling the type conversion in a centralized manner and it's a nice interface for comparing the service properties instead of having to do it in multiple places. 
I have modified and added a little bit to the dataclass you provided here, but I feel the way you mentioned here is definitely the way to go, can you please let me know if the changes I made are acceptable.

j-ororke committed:
Updating TC_CADMIN_1_5 test module fix:
- Updated to using dataclass to make code cleaner, thank you Andrei!

j-ororke commented on src/python_testing/TC_CADMIN_1_5.py:
```diff
@@ -44,13 +45,59 @@
+                    return service
+
+            # Log what we found for debugging purposes
+            if services:
```
This does seem better, I have removed this if check in the code here now

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating TC_CADMIN_1_5 test module:
- Actually removing if check here now

j-ororke committed:
Reverting value for expected_cm_value

raul-marquez-csa reviewed (approved):
LGTM

andy31415 commented on src/python_testing/TC_CADMIN_1_5.py:
```diff
@@ -44,13 +47,63 @@
+        d: Optional[int] = None
+
+        def __post_init__(self):
+            self.cm = int(self.service.txt_record.get('CM', None))
```
I think you should do some try/except because if you do miss CM you will have:

```
>>> int(None)                                                                                                                                                                     Traceback (most recent call last):                                                                                                                                              
File "<stdin>", line 1, in <module>                                                                                                                                            TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```

j-ororke committed:
Updating TC_CADMIN_1_5 test module:
- Added in if check with try and except block to catch and handle if CM or D values get set as None.

restyled-commits committed:
Restyled by autopep8
